[processor/tailsampling] Initial support for early decisions

csmarchbanks opened the pull request:
#### Description

This change adds a new optional interface that samplers can implement, `EarlyEvaluator`. If a sampler implements the `EarlyEvaluator` interface and the `EarlyDecisions` configuration is turned on then for each batch of traces received TSP will try to see if any spans from that batch will cause the trace to be sampled. If so, then we can send all the trace data we have along and remove it from memory. It is also possible that all policies return `NotSampled`, at which point TSP will drop the data.

In this initial implementation drop policies are not supported. In addition, the deprecated top level `invert_match` policies are also not supported and will run into issues if still being used. 

#### Link to tracking issue
Part of #43876 

#### Testing
I have added some tests and also ran this in some collectors in our infrastructure.

#### Documentation
TODO: I need to add more documentation than currently present

csmarchbanks committed:
Implement early decisions

Allow the processor to run early evaluations at batch ingest time and
make sampling decisions based on the result. This will drop or forward
traces before the entire batch is ready to reduce the memory required
keeping all trace data in memory until the end of the decision wait.

To begin with only implement "basic" samplers, those that do not support
invert, or nest other policies. This will still provide some gains while
keeping this change manageable.

Drop specifically will take more work to implement as we cannot make an
early Sampled decision until all drop policies have been evaluated which
will require some state to be maintained.

csmarchbanks committed:
Add parent span to sampling TraceData

csmarchbanks committed:
Clarify that allData should not be used on each request

yvrhdn commented on processor/tailsamplingprocessor/processor.go:
```diff
@@ -734,6 +790,71 @@ func (tsp *tailSamplingSpanProcessor) processTrace(id pcommon.TraceID, rss ptrac
+		earlyEval, ok := p.evaluator.(samplingpolicy.EarlyEvaluator)
+		if !ok {
+			continue
+		}
```
we can avoid the cast by using `p.earlyEvaluator` I think?


```suggestion
		earlyEval := p.earlyEvaluator
		if earlyEval == nil {
			continue
		}
```

yvrhdn commented on processor/tailsamplingprocessor/processor.go:
```diff
@@ -734,6 +790,71 @@ func (tsp *tailSamplingSpanProcessor) processTrace(id pcommon.TraceID, rss ptrac
+		decision, err := earlyEval.EarlyEvaluate(tsp.ctx, id, currentSpans, trace)
+		if err != nil {
+			tsp.telemetry.ProcessorTailSamplingSamplingPolicyEvaluationError.Add(tsp.ctx, 1)
+			tsp.logger.Debug("Sampling policy error", zap.Error(err))
```
Nit
```suggestion
			tsp.logger.Debug("Sampling policy error on early evaluation", zap.Error(err))
```

yvrhdn reviewed (commented):
I like the new interface! I think it's a good balance between giving policies the opportunity to decide fast without making it too heavy. Left a few minor comments.

Logiraptor commented on processor/tailsamplingprocessor/internal/sampling/latency.go:
```diff
@@ -37,10 +40,18 @@ func (l *latency) Evaluate(_ context.Context, _ pcommon.TraceID, traceData *samp
+	return hasSpanWithCondition(batches, l.condition()), nil
+}
+
+func (l *latency) EarlyEvaluate(_ context.Context, _ pcommon.TraceID, batch ptrace.ResourceSpans, _ *samplingpolicy.TraceData) (samplingpolicy.Decision, error) {
```
I think this only works correctly if the upper bound is unset, otherwise we can't know whether the trace will get longer later

Logiraptor reviewed (commented):
Sending a partial review while sitting in a hotel 😅, don't block on me while I'm OOO!

csmarchbanks committed:
Fix latency upper threshold behavior

csmarchbanks committed:
Fix review feedback

csmarchbanks committed:
Add EarlyEvaluate to the Evaluator interface

Combine the two interfaces previously created to simplify the code and
avoid as many casts everywhere. A sampler that does not support early
decisions can just return samplingpolicy.Unspecified.

csmarchbanks committed:
Fix and test NotSampled early decision case

csmarchbanks committed:
Fix lint errors

csmarchbanks committed:
Fix lint failures post rebase

csmarchbanks committed:
Add more early vs normal attributes for debugging

csmarchbanks committed:
Add changelog entry

csmarchbanks commented:
Closing this one for now as https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/44878 looks to have more impact in our environments. I may come back to it in the future.
